slam_toolbox/src/loop_closure_assistant.cpp

```diff
@@ -125,48 +125,93 @@ void LoopClosureAssistant::publishGraph()
 /*****************************************************************************/
 {
   interactive_server_->clear();
-  std::unordered_map<int, Eigen::Vector3d>* graph = solver_->getGraph();
+  karto::MapperGraph * graph = mapper_->GetGraph();
 
-  if (graph->size() == 0)
+  if (!graph || graph->GetVertices().empty())
   {
     return;
   }
 
-  ROS_DEBUG("Graph size: %i",(int)graph->size());
   bool interactive_mode = false;
   {
     boost::mutex::scoped_lock lock(interactive_mutex_);
     interactive_mode = interactive_mode_;
   }
 
-  visualization_msgs::MarkerArray marray;
-  visualization_msgs::Marker m = vis_utils::toMarker(map_frame_,
-    "slam_toolbox", 0.1);
+  // schedule all existing markers for removal
+  for (auto it = markers_.begin(); it != markers_.end(); ++it) {
+    it->second.action = visualization_msgs::Marker::DELETE;
+  }
+
+  visualization_msgs::Marker vertex_marker =
+    vis_utils::toVertexMarker(map_frame_, "slam_toolbox", 0.1);
 
-  for (ConstGraphIterator it = graph->begin(); it != graph->end(); ++it)
+  for (const auto & vertices_per_sensor : graph->GetVertices())
   {
-    m.id = it->first + 1;
-    m.pose.position.x = it->second(0);
-    m.pose.position.y = it->second(1);
+    for (const auto & id_and_vertex : vertices_per_sensor.second)
+    {
+      karto::LocalizedRangeScan * scan = id_and_vertex.second->GetObject();
+
+      vertex_marker.id = scan->GetUniqueId() + 1;
```

Comment:
+1? Its been a bit since I've worked in this part of the code, why this?

Reply:
That's preexisting code. Commit history doesn't shed much light. We can remove it.

Reply:
Lets keep it if its preexisting, that's fine.